server/jamovi/server/dataset/duckdataset.py

```diff
@@ -172,32 +216,206 @@ def __iter__(self) -> Iterator[DuckColumn]:
             yield column
 
-    def set_value(self, row: int, column: int, value, initing=False):
+    def set_value(
+        self,
+        row: int,
+        column: int | DuckColumn,
+        raw_value: int | float | str,
+        initing=False,
+    ):
         """set a value in the data set"""
-        col = self._columns_by_index[column]
 
-        if isinstance(value, str) and col.data_type is DataType.TEXT and col.has_levels:
-            try:
-                raw = col.get_value_for_label(value)
-            except KeyError:
-                raw = self.column_add_level(col, value)
-            value = raw
+        if isinstance(column, int):
+            column = self._columns_by_index[column]
 
+        if column.data_type is DataType.DECIMAL:
+            assert isinstance(raw_value, Number)
+        elif (
+            column.data_type is DataType.TEXT and column.measure_type is MeasureType.ID
+        ):
+            assert isinstance(raw_value, str)
+        else:
+            assert isinstance(raw_value, int)
+
+        value = raw_value
+        level_to_remove: int | None = None
+
+        if column.has_levels:
+            if not isinstance(raw_value, int):
+                raise ValueError
+            if column.data_type is DataType.TEXT:
+                if raw_value == -2147483648:
+                    value = ""
+                else:
+                    value = column.get_label(raw_value)
+
+            if not initing:
+                old_value = self.get_value(row, column)
+                if value == old_value:
+                    return
+
+            row_filtered = self.is_row_filtered(row)
+
+            if not initing and not is_missing(old_value):
+                for index, level in enumerate(column.dlevels):
+                    if column.data_type is DataType.TEXT:
+                        level_value = level.import_value
+                    else:
+                        level_value = level.value
+
+                    if old_value == level_value:
+                        # decrement count for old value
+                        level.count -= 1
+                        if level.count == 0:
+                            # delay removal until after the new value is set
+                            level_to_remove = index
+                        if not row_filtered:
+                            level.count_ex_filtered -= 1
+                        break
+                else:
+                    logger.warning("Shouldn't get here")
+
+            for level in column.dlevels:
+                if column.data_type is DataType.TEXT:
+                    level_value = level.import_value
+                else:
+                    level_value = level.value
+                if value == level_value:
+                    # increment count for new value
+                    level.count += 1
+                    if not row_filtered:
+                        level.count_ex_filtered += 1
```

Comment:
Is there a way to factor out some parts of this very nested piece of code? It's not very easy to follow.